[GTK][WPE] Wrong vector index calculated in Damage::unite()

carlosgcampos opened the pull request:
#### ac29a5c3090553a6fa69562670f267888d6424d5
<pre>
[GTK][WPE] Wrong vector index calculated in Damage::unite()
<a href="https://bugs.webkit.org/show_bug.cgi?id=289955">https://bugs.webkit.org/show_bug.cgi?id=289955</a>

Reviewed by Alejandro G. Castro.

The problem is that we are using ceiledIntPoint() resulting in cell
indices starting at 1, we should use flooredIntPoint instead to get the
right cell indices starting at 0.
This patch also includes other improvements and cleanups to Damage and a
small change in behavior, now we start doing the unite when adding a new
rectangle that would increase the vector size over the maximum instead
of when the maximum is reached.
The patch also includes a new unit test.

* LayoutTests/platform/glib/damage/storage-and-simplification-algorithm.html:
* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::Damage):
(WebCore::Damage::bounds const):
(WebCore::Damage::resize):
(WebCore::Damage::add):
(WebCore::Damage::uniteExistingRects):
(WebCore::Damage::tileIndexForRect const):
(WebCore::Damage::unite):
(WebCore::Damage::invalidate): Deleted.
(WebCore::Damage::shouldUnite const): Deleted.
* Tools/TestWebKitAPI/PlatformGTK.cmake:
* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp: Added.
(TestWebKitAPI::TEST(Damage, Basics)):
(TestWebKitAPI::TEST(Damage, AddRect)):
(TestWebKitAPI::TEST(Damage, AddDamage)):
(TestWebKitAPI::TEST(Damage, Unite)):

Canonical link: <a href="https://commits.webkit.org/292360@main">https://commits.webkit.org/292360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ff312280d608722bc871ab9247f0144d446fbeb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95592 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15194 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100641 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46097 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97636 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15480 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23631 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72905 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30171 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98595 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86350 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53237 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11311 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4063 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45433 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81500 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4182 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102676 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22642 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16539 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81951 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22894 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82364 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81302 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25885 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3354 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15986 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15414 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22610 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27767 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22269 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25745 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24011 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
```diff
@@ -0,0 +1,304 @@
+    EXPECT_EQ(damage.bounds().width(), 400);
+    EXPECT_EQ(damage.bounds().height(), 400);
+
+    // Adding a rect containg the bounds makes it the only rect.
```
```suggestion
    // Adding a rect containing the bounds makes it the only rect.
```

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -64,23 +72,29 @@ class Damage {
+
+        m_size = size;
+        m_gridSize = { std::max(1, m_size.width() / m_tileSize.width()), std::max(1, m_size.height() / m_tileSize.height()) };
+        m_rects.clear();
```
I'd rather expect the stored rects to be re-inserted in this case - alternatively (if used only during creation) perhaps we can make the `resize()` `private`?

carlosgcampos replied:
I don't think it's actually expected to resize a Damage after rectangles are added. The idea is to add a constructor taking the size, but it's not always possible, for example a class having a Damage as member might not know the size at construction time.

Scony replied:
right - sounds reasonable

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -25,7 +25,7 @@
 #pragma once
 
-#if ENABLE(DAMAGE_TRACKING)
+#if PLATFORM(GTK) || PLATFORM(WPE)
```
isn't it possible to pass `DAMAGE_TRACKING` to tests?

carlosgcampos replied:
Yes, but I changed the guards in Damage too, since I plan to use the class for other things not depending on DAMAGE_TRACKING being enabled.

Scony replied:
good point

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -44,7 +44,15 @@ class Damage {
+        : m_tileSize(s_tileSize, s_tileSize)
+    {
+        // FIXME: Add a constructor to pass the size.
+        resize({ 2048, 1024 });
```
shouldn't we have 2048x1024 as a `constexpr`, too? Or you're expecting this is a very short-lived FIXME?

carlosgcampos replied:
Yes, this is temporary.

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -91,82 +105,96 @@ class Damage {
         }
 
-        if (m_rects.size() == 1 && m_minimumBoundingRectangle.contains(rect))
+        auto rectsCount = m_rects.size();
```
could be `const`

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -91,82 +105,96 @@ class Damage {
 
         if (rect.contains(m_minimumBoundingRectangle)) {
-            m_rects = { rect };
+            m_rects.clear();
+            m_rects.append(rect);
```
`m_rects = { rect };` seemed clear enough, out of curiosity, is there a reason you prefer clear + append?

carlosgcampos replied:
Yes, maybe this is already optimized by the compiler, but since the Vector has one stack allocated item, in this case clearing the existing one and adding the rect seems cheaper than creating a new instance.

spenap commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -91,82 +105,96 @@ class Damage {
-        for (auto& rect : m_rects)
-            rect = { };
+        Rects rectsCopy(m_rects.size());
+        m_rects.swap(rectsCopy);
```
nice

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -91,82 +105,96 @@ class Damage {
             unite(rect);
     }
 
+    size_t tileIndexForRect(const IntRect& rect) const
```
perhaps `inline` ?

carlosgcampos replied:
I'm always confused by that, but aren't all functions defined in the header always inline?

spenap replied:
it's tricky. If they are small and the compiler feels like it, yes, but they might not. Unless going with `ALWAYS_INLINE`, which I guess expands to some pragma, the regular `inline` keyword is just a hint

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -91,82 +105,96 @@ class Damage {
-        const auto rectCenter = rect.center();
-        const auto rectCell = ceiledIntPoint(FloatPoint { static_cast<float>(rectCenter.x()) / tileSize.width(), static_cast<float>(rectCenter.y()) / tileSize.height() });
-        const size_t index = std::clamp(rectCell.x(), 0, static_cast<int>(s_gridCols - 1)) + std::clamp(rectCell.y(), 0, static_cast<int>(s_gridRows - 1)) * s_gridCols;
+        auto index = tileIndexForRect(rect);
```
could be `const`

Scony commented on Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
```diff
@@ -0,0 +1,304 @@
+    EXPECT_EQ(damage.bounds().width(), 200);
+    EXPECT_EQ(damage.bounds().height(), 200);
+
+    // When there's only once rect adding a rect already contained
```
```suggestion
    // When there's only one rect, adding a rect already contained
```

alexgcastro reviewed (approved):
LGTM!

carlosgcampos committed:
[GTK][WPE] Wrong vector index calculated in Damage::unite()
https://bugs.webkit.org/show_bug.cgi?id=289955

Reviewed by Alejandro G. Castro.

The problem is that we are using ceiledIntPoint() resulting in cell
indices starting at 1, we should use flooredIntPoint instead to get the
right cell indices starting at 0.
This patch also includes other improvements and cleanups to Damage and a
small change in behavior, now we start doing the unite when adding a new
rectangle that would increase the vector size over the maximum instead
of when the maximum is reached.
The patch also includes a new unit test.

* LayoutTests/platform/glib/damage/storage-and-simplification-algorithm.html:
* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::Damage):
(WebCore::Damage::bounds const):
(WebCore::Damage::resize):
(WebCore::Damage::add):
(WebCore::Damage::uniteExistingRects):
(WebCore::Damage::tileIndexForRect const):
(WebCore::Damage::unite):
(WebCore::Damage::invalidate): Deleted.
(WebCore::Damage::shouldUnite const): Deleted.
* Tools/TestWebKitAPI/PlatformGTK.cmake:
* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp: Added.
(TestWebKitAPI::TEST(Damage, Basics)):
(TestWebKitAPI::TEST(Damage, AddRect)):
(TestWebKitAPI::TEST(Damage, AddDamage)):
(TestWebKitAPI::TEST(Damage, Unite)):

Canonical link: https://commits.webkit.org/292360@main
